### ISSUE TITLE
Ekaterina Mamina: Behavior - Strategy - Implementation, 3 weeks

### ISSUE THREAD

**Issue by katsmamina** [HackYourFutureBelgium/class-15#240]
> [module repository](https://github.com/katsmamina/behavior-strategy-implementation)

## Study Plan
- [ ] 🥚 ./implicit-return
- [ ] 🥚 ./array-methods
- [ ] 🥚 ./ecmascript-modules (week 3)
- [ ] 🐣 ./hoisting
- [ ] 🐣 ./linting (week 3)
- [ ] 🐥 ./throw-and-catch
- [ ] 🐥 ./prototype-delegation

- [ ] 🥚 ./about-testing
- [ ] 🐣 ./parsons-problems
- [ ] 🐣 ./function-design
- [ ] 🐣 ./fuzz-testing

- [ ] 🥚,🐣,🐥 Solution Write-Ups 
- [ ] 🐣 document-and-pass
- [ ] 🐣 practice-code-review (week 3 group exercises)

**Comment by katsmamina** [comment]
## Week 1
- [ ] started the exercises

### Check-In

#### I Need Help With:
When I'm trying to run testing exercises from the repo in VSCode, it gives me error (such as 'describe' is not defined). 
Also, when I was trying to look into exercises in the debugger through `study-lenses`, it switches to another file , `describe-it.js`. I tried to look for solution in google, and found information that some `mocha` should be installed, is this correct?   
I remember we discussed the issue with Samir, but couldn't find back the answer. 

#### What went well?
Studying with classmates, analyzing code at edabit and running tests, started some exercises from the repo 

#### What went less well?
working in VSCode not in study lenses

#### Lessons Learned
difference between behavior, strategy and implementation 
the same task can be solved through very different code, it's beautiful
new syntax

### Sunday Prep Work
doing some more exercises from edabit

**Comment by colevandersWands** [comment]
> When I'm trying to run testing exercises from the repo in VSCode, it gives me error (such as 'describe' is not defined).

Are you using the normal `node file-name.js` to run the tested files?  if so that's your problem, you will need to use the debugger or `npm run test -- file-name.js`.  There's instructions for this in the README.  
If that doesn't work then it's most likely because you have an older version of the package.json & launch.json with a mistake in it. You can pull changes from the repo to fix that

> Also, when I was trying to look into exercises in the debugger through study-lenses, it switches to another file , describe-it.js. 

This is normal. `describe`, `it` and `expect` are not built into JS so the debugger will skip to the file where they are defined when you step through your tests.  You can use breakpoints to jump out of it, or the "Step Over" button to avoid going into it (https://javascript.info/debugging-chrome)

**Comment by katsmamina** [comment]
thanks @colevandersWands ! 

> Are you using the normal `node file-name.js` to run the tested files? if so that's your problem, you will need to use the debugger or `npm run test -- file-name.js`. There's instructions for this in the README.
I was trying to run some `file-name.test.js` , with the ' with tests' option. And the debug console is always telling me 'Process exited with code 1',  then sending me to the `error-handler.js` file. 
I can't even go through steps even after putting breakpoints! 
However, when I try to run the code through console as described in the README, it works, and I can see which tests were passed. 

> If that doesn't work then it's most likely because you have an older version of the package.json & launch.json with a mistake in it. You can pull changes from the repo to fix that
I pulled the changes but it didn't help :( 

> > Also, when I was trying to look into exercises in the debugger through study-lenses, it switches to another file , describe-it.js.
> This is normal. `describe`, `it` and `expect` are not built into JS so the debugger will skip to the file where they are defined when you step through your tests. You can use breakpoints to jump out of it, or the "Step Over" button to avoid going into it (https://javascript.info/debugging-chrome)
In the beginning, I used the 'step over' button but it didn't give me deep understanding what's going on, it just went to showing results in the console. So I tried to use 'Step' instead.
it's still unclear, where do these additional files come from ? should they be installed separately, in general? that was actually my question.

**Comment by colevandersWands** [comment]
> I was trying to run some file-name.test.js , with the ' with tests' option.
> I pulled the changes but it didn't help :(

can you push your repo as it is on your computer?  I can take a look on github and send you a PR if it's a problem with one of the files

**Comment by katsmamina** [comment]
Hi Evan, 
I updated the repo , could you take a quick look please?
by the way, I can't also run tests in the sandbox (for exercises on edabit), as VSCode says that there's no such file or directory. 
it seems like understanding VSC is harder than learning code features! :))

**Comment by colevandersWands** [comment]
Everything in the repo seems to be good, so it's most likely something you just need to practice.  VSCode definitely takes some practice!  But it's also not priority.  if you're able to run the tests in the browser then you're good to go for now.

> I can't also run tests in the sandbox (for exercises on edabit), as VSCode says that there's no such file or directory.

try updating this repo as well, there was the same mistake in `/.vscode/launch.json` and `package.json`.  If it's not that then you're likely not passing in a correct path after `npm run test -- `

**Comment by katsmamina** [comment]
It seems that I did all that was needed but still struggle with VSC, so switched to debugger + `study-lenses`.

In general, Evan, what would you recommend - to try to understand how VSCode works now, or leave it for later? Will we have some more time to practice it in the future? 

Maybe it would be possible to organize a call with you next week, to understand better how the program works? I don't always understand what is needed, and it would be much more efficient with screen sharing. Also, it seems that I am not the only one struggling with VSC, so maybe it would be useful for classmates as well.

**Comment by colevandersWands** [comment]
We can go over this in class tomorrow morning.  For now you don't need to use VSCode, it's good to slowly start practicing with it but you won't need to use it until Web Apps.  the more prepared you are then the better but it's also not worth losing too much study time now

**Comment by samirm00** [comment]
@colevandersWands @katsmamina  It is totally up to you! but I am not a fun of using VSC to debug or test  the Front-End , it is a good idea to knows how it works , but the browser or in our case `study-lenses` is  better and you will find a lot of senior developer do the same.  VSC is mainly for the Back-End , because it is not possible to debug it or test in the browser , that is why I ask you to do it in study-lenses.  what do you thing Evan ?

**Comment by katsmamina** [comment]
@samirm00 
Actually, my main concern was how I would write and test real code after studying at HYF without mastering VSC debugger, but it's a relief to hear that in front-end, browser might be enough 😂

**Comment by colevandersWands** [comment]
It's just what @samirm00 said - debug frontend code in the browser's debugger, and backend code in VSC.  all study-lenses does is put an editor right next to the browser's debugger for convenience.  

You'll see in Separation of Concerns when you start learning the DOM that everything you learned in Debugging transfers directly to real websites.

**Comment by katsmamina** [comment]
## Week 2

- [x] started the exercises

### Check-In

#### I Need Help With:

#### What went well?
I am concentrating on learning methods for function design

#### What went less well?
It seems like doing exercises takes more time than expected. It's good to know that we have one extra week for this module.
I was also trying to look into `extra--object-transformations`, and it seems quite clear in the examples yet less clear in exercises

#### Lessons Learned
- higher order functions for arrays
- a bit of object methods
- a bit of function design 

### Sunday Prep Work
- practice more function design
- understand object methods

**Comment by colevandersWands** [comment]
> I was also trying to look into extra--object-transformations, and it seems quite clear in the examples yet less clear in exercises

Don't worry so much about this, it's not something you'll need to use so often.  If you don't have much time you can always put this off until some day in the future when you need it

**Comment by katsmamina** [comment]
Thank you! by the way, a small question here.
While doing the exercise https://github.com/katsmamina/behavior-strategy-implementation/blob/master/function-design/exercises/medium/fizzbuzz-2.test.js ... I couldn't solve the problem until removing `++` on line 32, even though I tried almost everything I could think about. 
However, without  `++` the solution was found easily. Was there a mistake in the initial code, or should I have found a solution with the given code?

**Comment by colevandersWands** [comment]
Good catch, I think I'll replace it with `while (_) {` so the mistake is yours to make

**Comment by katsmamina** [comment]
## Week 3


### Check-In

#### I Need Help With:
So far it's ok

#### What went well?
I finished describing the solution that we started during the class. I wrote more tests, strategy, implementation and use cases.

#### What went less well?
I was distracted for a few days, so missed some time to work.

#### Lessons Learned
- Linting
- Formatting code
- Import-Export

### Sunday Prep Work
- hope to review solutions of other students
